libcontainer/cgroups/fs2/memory.go

```diff
@@ -104,10 +104,17 @@ func statMemory(dirPath string, stats *cgroups.Stats) error {
 
 	memoryUsage, err := getMemoryDataV2(dirPath, "")
 	if err != nil {
-		if errors.Is(err, unix.ENOENT) && dirPath == UnifiedMountpoint {
+		if errors.Is(err, unix.ENOENT) && dirPath == rootCgroupPath {
 			// The root cgroup does not have memory.{current,max}
 			// so emulate those using data from /proc/meminfo.
-			return statsFromMeminfo(stats)
+			if err := statsFromMeminfo(stats); err != nil {
+				return err
+			}
+			// cgroup v1 `usage_in_bytes` reports memory usage as rss (NR_ANON_MAPPED) + cache (NR_FILE_PAGES).
+			// cgroup v2 reports the same values as anon and file.
+			// sum those to report the same value as `usage_in_bytes` in v1 for consistency.
+			stats.MemoryStats.Usage.Usage = stats.MemoryStats.Stats["anon"] + stats.MemoryStats.Stats["file"]
```

Comment:
potentially we should add `swapcached` here for consistency with v1 as well...

Reply:
I don't think we should emulate the what cgroupv1 gives you for the cgroupv2 (and only for the root case too) -- in general in OCI we just give you what the kernel provides. I suspect emulating cgroupv2 more correctly (which is what you seem to be doing here? I need to look into how the kernel does the accounting...).

We have to do cgroupv1 -> cgroupv2 mapping for writing to cgroupfs (because otherwise old configs wouldn't work on newer systems) but faking cgroupv1 stats output for cgroupv2 seems particularly odd to me.

Also, this "fudging" should be done in the `statsFromMeminfo` code itself imho (where the rest of the calculation is).

 @kolyshkin wdyt?

Reply:
hmm. my original intention was to match `usage_in_bytes` which I think is NR_FILE_PAGES + NR_ANON_MAPPED (+ NR_SWAPCACHE) for the root cgroup. I'm not sure what the equivalent of `memory.current` would be for the root cgroup in v2 if not something like the v1 estimate? I admit I don't know specifics of how the per-cg usage counter charges.

I guess part of the problem is "usage" is poorly defined vs. specific zones/memory types? part of my read was that runc -> cadvisor -> k8s basically aligned on "usage" == NR_FILE_PAGES + NR_ANON_MAPPED, and "working set" == "usage" - "inactive_file". which has issues, but is at least consistent in v1 + v2 short of renamed statistics. 

I can definitely say `total - free` isn't working well, but welcome alternatives 😄

Reply:
@kolyshkin any thoughts on this one?

Reply:
As described in https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/cgroup-v1/memory.rst?plain=1#L625-L633 , I think maybe we should add `swapcached` here.
So if we want to merge this, we should make a decision for this question.

Reply:
I agree with adding swapcached here. One reservation I had -- `total - free` for swap also does not equal `swapcached` for the root, but if you look at specific slices, `memory.swap.current` for a given cgroup doesn't equal `memory.stat.swapcached` anyway. I suppose it makes sense -- swap area can be  used or not free in different ways (?), part of the swap space may not be writable (not the case for my demo however).

we also deliberately add memory + swap usage to match v1 later - 

```go
// As cgroup v1 reports SwapUsage values as mem+swap combined,
	// while in cgroup v2 swap values do not include memory,
	// report combined mem+swap for v1 compatibility.
	swapUsage.Usage += memoryUsage.Usage
	if swapUsage.Limit != math.MaxUint64 {
		swapUsage.Limit += memoryUsage.Limit
	}
	stats.MemoryStats.SwapUsage = swapUsage
```

should potentially make the same adjustment

Reply:
added swapcached, but what do you think about the swap adjustment?

I think we should add memory usage for consistency, e.g. you can see consumers expect it today https://github.com/google/cadvisor/blob/8164b38067246b36c773204f154604e2a1c962dc/container/libcontainer/handler.go#L808

but i'm not sure if we should stick with `total - free` for swap usage, use root swapcached, or something else?

Reply:
> should potentially make the same adjustment

I think so, maybe this is a bug before.
When we fall though to use `statsFromMeminfo` to calc node's memory data, we return in here, so `report combined mem+swap for v1 compatibility` is missing.
Please see:
https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/fs2/memory.go#L105-L126

@kolyshkin Am I right?

Reply:
think that might be better?

notably, I think we want nr_swap_pages from the kernel, but we only get proxies to that - 
https://github.com/torvalds/linux/blob/5d0c230f1de8c7515b6567d9afba1f196fb4e2f4/mm/memcontrol.c#L3677-L3680
https://github.com/torvalds/linux/blob/5d0c230f1de8c7515b6567d9afba1f196fb4e2f4/mm/swapfile.c#L3257-L3258

wonder if we have the same "overcounting" for swap or not